### PR TITLE
test(ci): test on windows and osx

### LINKED ISSUE #1648: [v3] Test environments and CI
We just merged #1592 which should make managing our various test environments much easier going forward. This ticket lays out a number of small tasks that can be taken to modernize our test environments and GitHub Actions Workflows. I'm sure I'm missing a few things here so feel free to comment / suggest alternatives.

### Hatch related improvements

- [x] Define dev environments in pyproject.toml using [Hatch environments](https://hatch.pypa.io/1.9/environment/)
  - Move `requirements_dev_*.txt` defining separate environments for "minimal", "dev". These should include parameterizations over python versions (3.10 and 3.11). 
  - Use these environments in our GitHub actions
- [x] Split out "minimal" environment into two separate environments
  - Minimum supported versions of all dependencies (see #1616)
  - Minimal dependencies (latest versions)
- [x] Create "docs" environment so developers can simply run: `hatch run docs:serve` 

### GitHub Actions related improvements

- [x] Explore removing the need to install a conda environment, instead relying on pure hatch environments
- [ ] Explore combining the "windows", "minimal", and "python-package" workflows into a single workflow
- [x] Run Mypy checks in each matrix run in addition to our pre-commit script

### PR DESCRIPTION
**Pull request by jhamman**
This PR adds CI matrix entries for OSX and Windows. 

Closes #1648 

Notes:

There are some interesting failures here.

~- [ ] path handling in local store~
- [x] async timeout error in osx

### CODE REVIEW

**Comment by martindurant** [comment]
> tweak path handling in the LocalStore

fsspec has a decent amount of work for this which you can reuse. Actually, could RemoteStore write to local too to avoid the duplication, or would that be unnecessarily complicated (given the async/sync difference in particular).

**Comment by jhamman** [comment]
> fsspec has a decent amount of work for this which you can reuse.

Can you provide some pointers?

> Actually, could RemoteStore write to local too to avoid the duplication, or would that be unnecessarily complicated (given the async/sync difference in particular).

I don't think this will help us because the `RemoteStore` requires an Async Filesystem which the fsspec local filesystem is not.

**Comment by martindurant** [comment]
fsspec calls https://github.com/fsspec/filesystem_spec/blob/master/fsspec/implementations/local.py#L284 as early as possible in localFS, since posix paths always work on windows in python (but not the other way around).

File: src/zarr/storage/local.py [hunk 1]
```diff
@@ -219,17 +219,17 @@ async def exists(self, key: str) -> bool:
...
+        to_strip = self.root.as_posix() + "/"
         for p in list(self.root.rglob("*")):
             if p.is_file():
-                yield str(p).replace(to_strip, "")
+                yield p.as_posix().replace(to_strip, "")
 
     async def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
         # docstring inherited
-        to_strip = os.path.join(str(self.root / prefix))
+        to_strip = (self.root / prefix).as_posix() + "/"  # TODO: fixme in 2430
```
Comment (jhamman) [inline src/zarr/storage/local.py:229]: see https://github.com/zarr-developers/zarr-python/issues/2438

Comment (jhamman) [inline src/zarr/storage/local.py:229]:
```suggestion
        to_strip = (self.root / prefix).as_posix() + "/"  # TODO: fixme in https://github.com/zarr-developers/zarr-python/issues/2438
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -16,14 +16,32 @@ concurrency:
...
+            dependency-set: 'optional'
+            os: 'macos-latest'
+          # - python-version: '3.11'
+          #   numpy-version: '1.25'
+          #   dependency-set: 'optional'
+          #   os: 'windows-latest'
+          # - python-version: '3.13'
+          #   numpy-version: '2.1'
+          #   dependency-set: 'optional'
+          #   os: 'windows-latest'
```
Comment (jhamman) [inline .github/workflows/test.yml:43]: I've opened a dedicated ticket for this: https://github.com/zarr-developers/zarr-python/issues/2438

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -16,14 +16,32 @@ concurrency:
...
+        include:
+          - python-version: '3.11'
+            numpy-version: '1.25'
+            dependency-set: 'optional'
+            os: 'macos-latest'
+          - python-version: '3.13'
+            numpy-version: '2.1'
+            dependency-set: 'optional'
+            os: 'macos-latest'
+          # - python-version: '3.11'
```
Comment (jhamman) [inline .github/workflows/test.yml:36]:
```suggestion
          # https://github.com/zarr-developers/zarr-python/issues/2438
          # - python-version: '3.11'
```

**Review by dcherian — APPROVED** [review]

**Comment by jhamman** [comment]
pre-commit.ci autofix

**Pull request merged**